### PR TITLE
[#IP-274] Implement opt-in/opt-out rules in `GetLimitedProfile` and `GetLimitedProfileByPOST` endpoint

### PR DESCRIPTION
**Pull request by balanza**
This is a rewrite of https://github.com/pagopa/io-functions-services/pull/136 in order to have a simpler implementation

#### List of Changes

* handle `AUTO` and `MANUAL` modes to determine if a Service can send messages to a Citizen in https://github.com/pagopa/io-functions-services/pull/141/commits/1067f58e74ee656ffa236255a5bf56d7cc888536
  * and tests in https://github.com/pagopa/io-functions-services/pull/141/commits/87d49556ef855f59ec3b454dceceded4ec487709 and https://github.com/pagopa/io-functions-services/pull/141/commits/3a06039f51184384a521cbb63ab0e7f39d62b8c1
* implement new business logic in exposed endpoints in https://github.com/pagopa/io-functions-services/pull/141/commits/718b8e754ceff430f541681f9cb2d8c576b9ee22

#### Motivation and Context

Services use `GetLimitedProfile` and `GetLimitedProfileByPOST` APIs to determine if they are allowed to send messages to a given Citizen. As there are now more ways in which a Citizen can manage its subscriptions, such APIs must handle new workflows. 

#### How Has This Been Tested?

Upgraded exisitng unit tests to match new model
Add unit test to cover the new profile-service implementation

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: utils/__tests__/profile.test.ts [hunk 1]
```diff
@@ -54,3 +116,90 @@ describe("retrievedProfileToLimitedProfile", () => {
...
+    ${"there are not preferences set for the service"} | ${"not allow"} | ${"MANUAL"} | ${some(aRetrievedProfileWithManualPreferences)}                                    | ${none}                                                            | ${false}
+    ${"there are not preferences set for the service"} | ${"allow"}     | ${"AUTO"}   | ${some(aRetrievedProfileWithAutoPreferences)}                                      | ${none}                                                            | ${true}
+    ${"the service is NOT in the blacklist"}           | ${"allow"}     | ${"LEGACY"} | ${some(withBlacklist(aRetrievedProfileWithLegacyPreferences, ["any-service-id"]))} | ${none}                                                            | ${true}
+    ${"has empty blacklist"}                           | ${"allow"}     | ${"LEGACY"} | ${some(withBlacklist(aRetrievedProfileWithLegacyPreferences, []))}                 | ${none}                                                            | ${true}
```
Comment (BurnedMarshal) [inline utils/__tests__/profile.test.ts:150]: We can add a test for blacklisted service in this suite (the last of this list)
```suggestion
    preferencesConfiguration                           | allowOrNot     | mode        | maybeProfile                                                                                              | maybePreference                                                    | expected
    ${"the inbox is enabled in the preferences"}       | ${"allow"}     | ${"MANUAL"} | ${some(aRetrievedProfileWithManualPreferences)}                                                           | ${some({ ...aRetrievedServicePreference, isInboxEnabled: true })}  | ${true}
    ${"the inbox is enabled in the preferences"}       | ${"allow"}     | ${"AUTO"}   | ${some(aRetrievedProfileWithAutoPreferences)}                                                             | ${some({ ...aRetrievedServicePreference, isInboxEnabled: true })}  | ${true}
    ${"the inbox is NOT enabled in the preferences"}   | ${"not allow"} | ${"MANUAL"} | ${some(aRetrievedProfileWithManualPreferences)}                                                           | ${some({ ...aRetrievedServicePreference, isInboxEnabled: false })} | ${false}
    ${"the inbox is NOT enabled in the preferences"}   | ${"not allow"} | ${"AUTO"}   | ${some(aRetrievedProfileWithAutoPreferences)}                                                             | ${some({ ...aRetrievedServicePreference, isInboxEnabled: false })} | ${false}
    ${"there are not preferences set for the service"} | ${"not allow"} | ${"MANUAL"} | ${some(aRetrievedProfileWithManualPreferences)}                                                           | ${none}                                                            | ${false}
    ${"there are not preferences set for the service"} | ${"allow"}     | ${"AUTO"}   | ${some(aRetrievedProfileWithAutoPreferences)}                                                             | ${none}                                                            | ${true}
    ${"the service is NOT in the blacklist"}           | ${"allow"}     | ${"LEGACY"} | ${some(withBlacklist(aRetrievedProfileWithLegacyPreferences, ["any-service-id"]))}                        | ${none}                                                            | ${true}
    ${"has empty blacklist"}                           | ${"allow"}     | ${"LEGACY"} | ${some(withBlacklist(aRetrievedProfileWithLegacyPreferences, []))}                                        | ${none}                                                            | ${true}
    ${"the service is in the blacklist"}               | ${"not allow"} | ${"LEGACY"} | ${some(withBlacklist(aRetrievedProfileWithLegacyPreferences, [anAzureUserAttributes.service.serviceId]))} | ${none}                                                            | ${false}
```

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

**Comment by balanza** [comment]
In https://github.com/pagopa/io-functions-services/pull/141/commits/95d0d36b8286eef80a8fad922b805de32942dfd1 I removed tests from http handlers and move them near the shared task.

This because they were actually duplicated. Case coverage remains the same.

File: utils/__tests__/arbitraries.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import { BlockedInboxOrChannelEnum } from "@pagopa/io-functions-commons/dist/generated/definitions/BlockedInboxOrChannel";
 import { NewMessage } from "@pagopa/io-functions-commons/dist/generated/definitions/NewMessage";
 import { PreferredLanguageEnum } from "@pagopa/io-functions-commons/dist/generated/definitions/PreferredLanguage";
+import { ServicesPreferencesModeEnum } from "@pagopa/io-functions-commons/dist/generated/definitions/ServicesPreferencesMode";
```
[the remark this answers is not in the document]
Comment (balanza) [reply utils/__tests__/arbitraries.ts]: fixed in https://github.com/pagopa/io-functions-services/pull/141/commits/95d0d36b8286eef80a8fad922b805de32942dfd1

File: utils/profile.ts [hunk 3]
```diff
@@ -15,40 +15,113 @@ import {
...
+): TaskEither<never, boolean> =>
+  taskEither.of(
     blockedInboxOrChannels === undefined ||
-    blockedInboxOrChannels[serviceId] === undefined ||
-    blockedInboxOrChannels[serviceId].indexOf(BlockedInboxOrChannelEnum.INBOX) <
-      0
+      blockedInboxOrChannels[serviceId] === undefined ||
+      blockedInboxOrChannels[serviceId].indexOf(
+        BlockedInboxOrChannelEnum.INBOX
+      ) < 0
```
Comment (BurnedMarshal) [inline utils/profile.ts:83]:
```suggestion
      !blockedInboxOrChannels[serviceId].includes(
        BlockedInboxOrChannelEnum.INBOX
      )
```

File: utils/profile.ts [hunk 4]
```diff
@@ -15,40 +15,113 @@ import {
...
+
+/**
+ * Whether the sender service is allowed to send
+ * messages to the user identified by this profile
+ * which servicesPreferencesSettings.mode is NOT LEGACY
+ */
+export const isSenderAllowed = (
+  servicesPreferencesModel: ServicesPreferencesModel,
+  serviceId: ServiceId,
+  profileId: FiscalCode,
```
Comment (BurnedMarshal) [inline utils/profile.ts:94]: We never use profileId as variable name for the user fiscal code. Can we use `fiscalCode` as usual?
```suggestion
  fiscalCode: FiscalCode,
```

**Review by BurnedMarshal — APPROVED** [review]
LGTM, only some stylistic consideration.

File: utils/__tests__/profile.test.ts [hunk 5]
```diff
@@ -54,3 +90,165 @@ describe("retrievedProfileToLimitedProfile", () => {
...
+    const { blockedInboxOrChannels = {}, ...restOfProfile } = p;
+    return {
+      ...restOfProfile,
+      blockedInboxOrChannels: {
+        ...blockedInboxOrChannels,
+        [serviceId]: [BlockedInboxOrChannelEnum.INBOX]
+      }
+    };
+  }, profile);
+
```
Comment (gquadrati) [inline utils/__tests__/profile.test.ts:133]: Another way without recreating profile 
```suggestion
// utility that adds a given set of serviceIds to the profile's inbox blacklist
const withBlacklist = (profile: RetrievedProfile, services = []) => {
  return {
    ...profile,
    blockedInboxOrChannels: services.reduce(
      (obj, serviceId) => ({
        ...obj,
        [serviceId]: BlockedInboxOrChannelEnum.INBOX
      }),
      {}
    )
  };
```

File: utils/profile.ts [hunk 6]
```diff
@@ -15,40 +15,113 @@ import {
...
+      | ServicesPreferencesModeEnum.AUTO
+      | ServicesPreferencesModeEnum.MANUAL;
+    readonly version: NonNegativeInteger;
+  }
+): TaskEither<CosmosErrors | IUnexpectedValue, boolean> =>
+  taskEither
+    .of<
+      CosmosErrors | IUnexpectedValue,
+      ReturnType<typeof makeServicesPreferencesDocumentId>
+    >(makeServicesPreferencesDocumentId(profileId, serviceId, version))
```
Comment (balanza) [inline utils/profile.ts:109]:
```suggestion
    >(makeServicesPreferencesDocumentId(fiscalCode, serviceId, version))
```

File: utils/profile.ts [hunk 7]
```diff
@@ -15,40 +15,113 @@ import {
...
+    .chain(docId => servicesPreferencesModel.find([docId, profileId]))
```
Comment (balanza) [inline utils/profile.ts:110]:
```suggestion
    .chain(docId => servicesPreferencesModel.find([docId, fiscalCode]))
````

**Pull request merged**